// js-to-ts-converter/dist/create-ts-ast-project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ts_simple_ast_1 = require("ts-simple-ast");
const fs = require("fs");
const Minimatch = require('minimatch').Minimatch;
const glob = require('glob-all');
/**
 * Creates a ts-simple-ast Project by including the source files under the given
 * `directory`.
 *
 * @param directory The absolute path to the directory of .js files to
 *   include.
 * @param options
 * @param options.indentationText The text used to indent new class property
 *   declarations.
 * @param options.excludePatterns Glob patterns to exclude files.
 */
function createTsAstProject(directory, options = {}) {
    const tsAstProject = new ts_simple_ast_1.default({
        manipulationSettings: {
            indentationText: options.indentationText || ts_simple_ast_1.IndentationText.Tab
        }
    });
    // Get all files, and then filter. Was using glob-all and passing all of the
    // globs to the utility, but it takes way too long on large projects because
    // it seems to read the file system multiple times - once for each pattern.
    let files = glob.sync(`${directory}/**/*.+(js|ts)`, {
        follow: true // follow symlinks
    });
    // First, filter out any path which includes node_modules. We don't want to
    // attempt to parse those as they may be ES5, and we also don't accidentally
    // want to write out into the node_modules folder
    const nodeModulesRegex = /[\\\/]node_modules[\\\/]/;
    files = files.filter((file) => !nodeModulesRegex.test(file));
    let includeMinimatches = createIncludeMinimatches(directory, options.includePatterns);
    let excludeMinimatches = createExcludeMinimatches(directory, options.excludePatterns);
    let includedFiles = files
        .filter((filePath) => {
        return includeMinimatches.some(minimatch => minimatch.match(filePath));
    })
        .filter((filePath) => {
        return !excludeMinimatches.some(minimatch => minimatch.match(filePath));
    })
        .filter((filePath) => fs.statSync(filePath).isFile()); // don't take directories
    includedFiles.forEach((filePath) => {
        tsAstProject.addExistingSourceFile(filePath);
    });
    return tsAstProject;
}
exports.createTsAstProject = createTsAstProject;
function createIncludeMinimatches(directory, includePatterns) {
    return (includePatterns || ['**/*.+(js|ts)'])
        .map(pattern => `${directory}/${pattern}`)
        .map(pattern => new Minimatch(pattern));
}
function createExcludeMinimatches(directory, excludePatterns) {
    return (excludePatterns || [])
        .map(pattern => `${directory}/${pattern}`)
        .map(pattern => new Minimatch(pattern));
}
